File: benchmarks/Java/src/json/JsonPureStringParser.java
/*******************************************************************************
 * Copyright (c) 2015 Stefan Marr
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 ******************************************************************************/
package json;

public final class JsonPureStringParser {

  private final String input;
  private int index;
  private int line;
  private int column;
  private String current;
  private String captureBuffer;
  private int captureStart;

  public JsonPureStringParser(final String string) {
    this.input = string;
    index = -1;
    line = 1;
    captureStart = -1;
    column = 0;
    current = null;
    captureBuffer = "";
  }

  public JsonValue parse() {
    read();
    skipWhiteSpace();
    JsonValue result = readValue();
    skipWhiteSpace();
    if (!isEndOfText()) {
      throw error("Unexpected character");
    }
    return result;
  }

  private JsonValue readValue() {
    switch(current) {
    case "n":
      return readNull();
    case "t":
      return readTrue();
    case "f":
      return readFalse();
    case "\"":
      return readString();
    case "[":
      return readArray();
    case "{":
      return readObject();
    case "-":
    case "0":
    case "1":
    case "2":
    case "3":
    case "4":
    case "5":
    case "6":
    case "7":
    case "8":
    case "9":
      return readNumber();
    default:
      throw expected("value");
    }
  }

  private JsonArray readArray() {
    read();
    JsonArray array = new JsonArray();
    skipWhiteSpace();
    if (readChar("]")) {
      return array;
    }
    do {
      skipWhiteSpace();
      array.add(readValue());
      skipWhiteSpace();
    } while (readChar(","));
    if (!readChar("]")) {
      throw expected("',' or ']'");
    }
    return array;
  }

  private JsonObject readObject() {
    read();
    JsonObject object = new JsonObject();
    skipWhiteSpace();
    if (readChar("}")) {
      return object;
    }
    do {
      skipWhiteSpace();
      String name = readName();
      skipWhiteSpace();
      if (!readChar(":")) {
        throw expected("':'");
      }
      skipWhiteSpace();
      object.add(name, readValue());
      skipWhiteSpace();
    } while (readChar(","));

    if (!readChar("}")) {
      throw expected("',' or '}'");
    }
    return object;
  }

  private String readName() {
    if (!current.equals("\"")) {
      throw expected("name");
    }
    return readStringInternal();
  }

  private JsonValue readNull() {
    read();
    readRequiredChar("u");
    readRequiredChar("l");
    readRequiredChar("l");
    return JsonLiteral.NULL;
  }

  private JsonValue readTrue() {
    read();
    readRequiredChar("r");
    readRequiredChar("u");
    readRequiredChar("e");
    return JsonLiteral.TRUE;
  }

  private JsonValue readFalse() {
    read();
    readRequiredChar("a");
    readRequiredChar("l");
    readRequiredChar("s");
    readRequiredChar("e");
    return JsonLiteral.FALSE;
  }

  private void readRequiredChar(final String ch) {
    if (!readChar(ch)) {
      throw expected("'" + ch + "'");
    }
  }

  private JsonValue readString() {
    return new JsonString(readStringInternal());
  }

  private String readStringInternal() {
    read();
    startCapture();
    while (!current.equals("\"")) {
      if (current.equals("\\")) {
        pauseCapture();
        readEscape();
        startCapture();
      } else {
        read();
      }
    }
    String string = endCapture();
    read();
    return string;
  }

  private void readEscape() {
    read();
    switch(current) {
    case "\"":
    case "/":
    case "\\":
      captureBuffer += current;
      break;
    case "b":
      captureBuffer += "\b";
      break;
    case "f":
      captureBuffer += "\f";
      break;
    case "n":
      captureBuffer += "\n";
      break;
    case "r":
      captureBuffer += "\r";
      break;
    case "t":
      captureBuffer += "\t";
      break;
    default:
      throw expected("valid escape sequence");
    }
    read();
  }

  private JsonValue readNumber() {
    startCapture();
    readChar("-");
    String firstDigit = current;
    if (!readDigit()) {
      throw expected("digit");
    }
    if (!firstDigit.equals("0")) {
      // Checkstyle: stop
      while (readDigit()) { }
     // Checkstyle: resume
    }
    readFraction();
    readExponent();
    return new JsonNumber(endCapture());
  }

  private boolean readFraction() {
    if (!readChar(".")) {
      return false;
    }
    if (!readDigit()) {
      throw expected("digit");
    }
    // Checkstyle: stop
    while (readDigit()) { }
    // Checkstyle: resume
    return true;
  }

  private boolean readExponent() {
    if (!readChar("e") && !readChar("E")) {
      return false;
    }
    if (!readChar("+")) {
      readChar("-");
    }
    if (!readDigit()) {
      throw expected("digit");
    }

    // Checkstyle: stop
    while (readDigit()) { }
    // Checkstyle: resume
    return true;
  }

  private boolean readChar(final String ch) {
    if (!current.equals(ch)) {
      return false;
    }
    read();
    return true;
  }

  private boolean readDigit() {
    if (!isDigit()) {
      return false;
    }
    read();
    return true;
  }

  private void skipWhiteSpace() {
    while (isWhiteSpace()) {
      read();
    }
  }

  private void read() {
    if ("\n".equals(current)) {
      line++;
      column = 0;
    }
    index++;
    if (index < input.length()) {
      current = input.substring(index, index + 1);
    } else {
      current = null;
    }
  }

  private void startCapture() {
    captureStart = index;
  }

  private void pauseCapture() {
    int end = current == null ? index : index - 1;
    captureBuffer += input.substring(captureStart, end + 1);
    captureStart = -1;
  }

  private String endCapture() {
    int end = current == null ? index : index - 1;
    String captured;
    if ("".equals(captureBuffer)) {
      captured = input.substring(captureStart, end + 1);
    } else {
      captureBuffer += input.substring(captureStart, end + 1);
      captured = captureBuffer;
      captureBuffer = "";
    }
    captureStart = -1;
    return captured;
  }

  private ParseException expected(final String expected) {
    if (isEndOfText()) {
      return error("Unexpected end of input");
    }
    return error("Expected " + expected);
  }

  private ParseException error(final String message) {
    return new ParseException(message, index, line, column - 1);
  }

  private boolean isWhiteSpace() {
    return " ".equals(current) || "\t".equals(current) || "\n".equals(current) || "\r".equals(current);
  }

  private boolean isDigit() {
    return "0".equals(current) ||
        "1".equals(current) ||
        "2".equals(current) ||
        "3".equals(current) ||
        "4".equals(current) ||
        "5".equals(current) ||
        "6".equals(current) ||
        "7".equals(current) ||
        "8".equals(current) ||
        "9".equals(current);
  }

  private boolean isEndOfText() {
    return current == null;
  }
}

File: benchmarks/C++/src/json/JsonPureStringParser.cpp
#include "JsonPureStringParser.h"

namespace json {

            shared_ptr<JsonValue> JsonPureStringParser::readValue() {
                if (_current == "n") {
                    return readNull();
                } else if (_current == "t") {
                    return readTrue();
                } else if (_current == "f") {
                    return readFalse();
                } else if (_current == "\"") {
                    return readString();
                } else if (_current == "[") {
                    return readArray();
                } else if (_current == "{") {
                    return readObject();
                }
                else if (_current == "-" || _current == "0" || _current == "1" || _current == "2" || _current == "3" || _current == "4"
                    || _current == "5" || _current == "6" || _current == "7" || _current == "8" || _current == "9") {
                    return readNumber();
                }
                else 
                    throw expected("value");
            }

            shared_ptr<JsonObject> JsonPureStringParser::readObject() {
                read();
                shared_ptr<JsonObject> object = make_shared<JsonObject>();
                skipWhiteSpace();
                if (readChar("}")) {
                    return object;
                }
                do {
                    skipWhiteSpace();
                    string name = readName();
                    skipWhiteSpace();
                    if (!readChar(":")) {
                        throw expected("':'");
                    }
                    skipWhiteSpace();

                    object->add(name, readValue());
                    skipWhiteSpace();
                } 
                while (readChar(","));
                if (!readChar("}")) {
                    throw expected("',' or '}'");
                }
                return object;
            }

            string JsonPureStringParser::readName() {
                if (_current != "\"") {
                    throw expected("name");
                }
                return readStringInternal();
            }

            shared_ptr<JsonArray> JsonPureStringParser::readArray() {
                read();
                shared_ptr<JsonArray> array = make_shared<JsonArray>();
                skipWhiteSpace();
                if (readChar("]")) {
                    return array;
                }
                do {
                    skipWhiteSpace();
                    array->add(readValue());
                    skipWhiteSpace();
                }  while (readChar(","));

                if (!readChar("]")) {
                        throw expected("',' or ']'");
                }
                return array;
            }

            shared_ptr<JsonValue> JsonPureStringParser::readNull() {
                read();
                readRequiredChar("u");
                readRequiredChar("l");
                readRequiredChar("l");
                return JsonLiteral::NNULL;
            }

            shared_ptr<JsonValue> JsonPureStringParser::readTrue() {
                read();
                readRequiredChar("r");
                readRequiredChar("u");
                readRequiredChar("e");
                return JsonLiteral::TRUE;
            }

            shared_ptr<JsonValue> JsonPureStringParser::readFalse() {
                read();
                readRequiredChar("a");
                readRequiredChar("l");
                readRequiredChar("s");
                readRequiredChar("e");
                return JsonLiteral::FALSE;
            }

            void JsonPureStringParser::readRequiredChar(string ch) {
                if (!readChar(ch)) {
                    throw expected("'" + ch + "'");
                }
            }

            shared_ptr<JsonValue> JsonPureStringParser::readString() {
                return make_shared<JsonString>(readStringInternal());
            }

            string JsonPureStringParser::readStringInternal() {
                read();
                startCapture();
                while (_current != "\"") {
                    if (_current == "\\") {
                        pauseCapture();
                        readEscape();
                        startCapture();
                    } else {
                        read();
                    }
                }
                string string = endCapture();
                read();
                return string;
            }

            void JsonPureStringParser::readEscape() {
                read();
                if (_current == "\"" || _current == "/" || _current == "\\") 
                    _captureBuffer += _current;
                else if (_current == "b")
                    _captureBuffer += "\b";
                else if (_current == "f")
                    _captureBuffer += "\f";
                else if (_current == "n")
                    _captureBuffer += "\n";
                else if (_current == "r")
                    _captureBuffer += "\r";
                else if (_current == "t")
                    _captureBuffer += "\t";
                else 
                    throw expected("valid escape sequence");
                read();
            }

            shared_ptr<JsonValue> JsonPureStringParser::readNumber() {
                startCapture();
                readChar("-");
                string firstDigit = _current;
                if (!readDigit()) {
                    throw expected("digit");
                }
                if (firstDigit != "0") {
                    while (readDigit()) { }
                }
                readFraction();
                readExponent();
                return make_shared<JsonNumber>(endCapture());
            }

            bool JsonPureStringParser::readFraction() {
                if (!readChar(".")) {
                    return false;
                }
                if (!readDigit()) {
                    throw expected("digit");
                }
                while (readDigit()) { }
                return true;
            }

            bool JsonPureStringParser::readExponent() {
                if (!readChar("e") && !readChar("E")) {
                    return false;
                }
                if (!readChar("+")) {
                    readChar("-");
                }
                if (!readDigit()) {
                    throw expected("digit");
                }

                while (readDigit()) { }
                return true;
            }

            bool JsonPureStringParser::readChar(string ch) {
                if (_current != ch) {
                    return false;
                }
                read();
                return true;
            }

            bool JsonPureStringParser::readDigit() {
                if (!isDigit()) {
                    return false;
                }
                read();
                return true;
            }

            void JsonPureStringParser::skipWhiteSpace() {
                while (isWhiteSpace()) {
                    read();
                }
            }

            void JsonPureStringParser::read() {
                if ("\n" == _current) {
                    _line++;
                    _column = 0;
                }
                _index++;
                if (_index < _input.length()) {
                    _current = _input.substr(_index, 1);
                } else {
                    _current = "";
                }
            }

            void JsonPureStringParser::startCapture() {
                _captureStart = _index;
            }

            void JsonPureStringParser::pauseCapture() {
                int _end = _current == "" ? _index : _index - 1;
                _captureBuffer += _input.substr(_captureStart, _end - _captureStart + 1);
                _captureStart = -1;
            }

            string JsonPureStringParser::endCapture() {
                int _end = _current == "" ? _index : _index - 1;
                string captured;
                if ("" == _captureBuffer) {
                    captured = _input.substr(_captureStart, _end - _captureStart + 1);
                } else {
                    _captureBuffer += _input.substr(_captureStart, _end - _captureStart + 1);
                    captured = _captureBuffer;
                    _captureBuffer = "";
                }
                _captureStart = -1;
                return captured;
            }

            ParseException JsonPureStringParser::expected(string expected) {
                if (isEndOfText()) {
                    return error("Unexpected end of input");
                }

                return error("Expected " + expected);
            }

            ParseException JsonPureStringParser::error(string message) {
                return ParseException(message, _index, _line, _column - 1);
            }

            bool JsonPureStringParser::isWhiteSpace() {
                return " " == _current || "\t" == _current || "\n" == _current || "\r" == _current;
            }


            bool JsonPureStringParser::isDigit() {
                return "0" == _current ||
                    "1" == _current ||
                    "2" == _current ||
                    "3" == _current ||
                    "4" == _current ||
                    "5" == _current ||
                    "6" == _current ||
                    "7" == _current ||
                    "8" == _current ||
                    "9" == _current;
            }

            bool JsonPureStringParser::isEndOfText() {
                return _current == "";
            }

            JsonPureStringParser::JsonPureStringParser(string string) {
                _input = string;
                _index = -1;
                _line = 1;
                _captureStart = -1;
                _column = 0;
                _current = "";
                _captureBuffer = "";
            }

            shared_ptr<JsonValue> JsonPureStringParser::parse() {
                read();
                skipWhiteSpace();
                shared_ptr<JsonValue> result = readValue();
                skipWhiteSpace();
                if (!isEndOfText()) {
                    throw Error("Unexpected character");
                }
                return result;
            }

}
